RecoLocalCalo/EcalRecProducers/plugins/alpaka/TimeComputationKernels.h

```diff
@@ -0,0 +1,1162 @@
+          if (ltx <= 8) {
+            sample_i = 0;
+            sample_j = 1 + ltx;
+          } else if (ltx <= 16) {
+            sample_i = 1;
+            sample_j = 2 + ltx - 9;
+          } else if (ltx <= 23) {
+            sample_i = 2;
+            sample_j = 3 + ltx - 17;
+          } else if (ltx <= 29) {
+            sample_i = 3;
+            sample_j = 4 + ltx - 24;
+          } else if (ltx <= 34) {
+            sample_i = 4;
+            sample_j = 5 + ltx - 30;
+          } else if (ltx <= 38) {
+            sample_i = 5;
+            sample_j = 6 + ltx - 35;
+          } else if (ltx <= 41) {
+            sample_i = 6;
+            sample_j = 7 + ltx - 39;
+          } else if (ltx <= 43) {
+            sample_i = 7;
+            sample_j = 8 + ltx - 42;
+          } else if (ltx <= 44) {
+            sample_i = 8;
+            sample_j = 9;
+          } else
+            assert(false);
+
+          auto const tx_i = ch_start + sample_i;
+          auto const tx_j = ch_start + sample_j;
+
+          //
+          // note, given the way we partition the block, with 45 threads per channel
+          // we will end up with inactive threads which need to be dragged along
+          // through the synching point
+          //
+          bool const condForUselessSamples = useless_sample_values[tx_i] || useless_sample_values[tx_j] ||
+                                             sample_values[tx_i] <= 1 || sample_values[tx_j] <= 1;
+
+          //
+          // see cpu implementation for explanation
+          //
+          ScalarType chi2 = std::numeric_limits<ScalarType>::max();
+          ScalarType tmax = 0;
+          ScalarType tmaxerr = 0;
+          shrTimeMax[idx.local] = 0;
+          shrTimeWgt[idx.local] = 0;
+
+          bool internalCondForSkipping1 = true;
+          bool internalCondForSkipping2 = true;
+          if (!condForUselessSamples) {
+            auto const rtmp = sample_values[tx_i] / sample_values[tx_j];
+            auto const invampl_i = 1. / sample_values[tx_i];
+            auto const relErr2_i = sample_value_errors[tx_i] * sample_value_errors[tx_i] * invampl_i * invampl_i;
+            auto const invampl_j = 1. / sample_values[tx_j];
+            auto const relErr2_j = sample_value_errors[tx_j] * sample_value_errors[tx_j] * invampl_j * invampl_j;
+            auto const err1 = rtmp * rtmp * (relErr2_i + relErr2_j);
+            auto err2 =
+                sample_value_errors[tx_j] * (sample_values[tx_i] - sample_values[tx_j]) * (invampl_j * invampl_j);
+            // TODO non-divergent branch for a block if each block has 1 channel
+            // otherwise non-divergent for groups of 45 threads
+            // at this point, pedestal_nums[ch] can be either 0, 1 or 2
+            if (pedestal_nums[ch] == 2)
+              err2 *= err2 * 0.5;
+            auto const err3 = (0.289 * 0.289) * (invampl_j * invampl_j);
+            auto const total_error = std::sqrt(err1 + err2 + err3);
```

Comment:
How are the results affected if these operations are done using `float` instead of `double` precision ?

Reply:
Can be checked later, see https://github.com/cms-sw/cmssw/issues/43551 .